org.eclipse.xtext/src/org/eclipse/xtext/build/IncrementalBuilder.java

```diff
@@ -358,6 +370,76 @@ protected void generate(Resource resource, BuildRequest request, Source2Generate
 			}
 		}
 
+		/**
+		 * @since 2.28
+		 */
+		protected boolean isLoadedFromStorage(Resource resource) {
+			return resource instanceof StorageAwareResource && ((StorageAwareResource) resource).isLoadedFromStorage();
+		}
+
+		/**
+		 * @since 2.28
+		 */
+		protected void installSourceLevelURIs(List<URI> uris) {
+			Set<URI> sourceLevelURIs = new HashSet<>(uris.size());
+			ResourceSet resourceSet = request.getResourceSet();
+			for (URI uri : uris) {
+				if (isSource(uri)) {
+					sourceLevelURIs.add(uri);
+					unloadResource(uri, this::isLoadedFromStorage);
+				}
+			}
+			SourceLevelURIsAdapter.setSourceLevelUrisWithoutCopy(resourceSet, sourceLevelURIs);
+		}
+
+		/**
+		 * Return true if the given uri must be loaded from source.
+		 * 
+		 * @since 2.28
+		 */
+		protected boolean isSource(URI uri) {
+			IResourceServiceProvider provider = context.getResourceServiceProvider(uri);
+			return provider instanceof IResourceServiceProviderExtension
+					&& ((IResourceServiceProviderExtension) provider).isSource(uri);
+		}
+		
+		/**
+		 * Deletes the binary models for the given set of URIs. Might be customized if
+		 * IResourceStorage supports deletion
+		 *
+		 * @param toBeDeleted
+		 *            set of URIs, must not be {@code null}
+		 * 
+		 * @since 2.28
+		 */
+		protected void deleteBinaryResources(final List<URI> toBeDeleted) {
```

Comment:
Wouldn't this need a fileSystemAccess?

Reply:
Yes, that is right, one would need a fileSystemAccess. I have assumed that customizations could create one, I did not do it in your codebase because at the moment one would need to call deleteBinaryResources there is no such fileSystemAccess and creating one if the method does not do anything, would not be necessary.
For the time being, I have removed the method, I will come back to it later.